### ISSUE TITLE
Make Transport a member of each Protocol impl, remove Transport from Protocol trait

### ISSUE THREAD

**Issue by gsingh93** [maximg/thrift#30]
A `Protocol` requires a `Transport` to be useful. Currently the transport is passed in as an argument to each trait method, but it would be more convenient to just make it a member of the `Protocol`. This would make implementing the server processor more convenient as well, as one can just pass around protocols instead of passing around protocols and transports.

Thoughts on this @maximg? This is what the C++ implementation does.

**Comment by gsingh93** [comment]
Thinking about this more, this would mean multiple instances of `Protocol` would own the same transport. If we did this, we'd have to use an `Rc` or an `Arc`.

**Comment by maximg** [comment]
I agree that we can model the protocol could own the transport, or several
protocols could talk to one transport wrapped in an Rc (probably not an Arc
though, unless we prevent the clients talking corrupting each other's
communications). So far the primary reason to keep the transport separately
was to avoid tracking the lifetimes and keep things simple.

I could imagine that as an alternative you could wrap the protocol and the
transport in a struct, similar to the way it is done in the generated
clients?

If you decide to go for ownership it will be good to see how much impact
the change will have on the generator.

**Comment by in-the-mood-for-mov** [comment]
IIRC the only reason I wrote it that way was that Rust makes mutability explicit unlike C++ and making the protocol own the transport would force us to make the protocol mutable, which felt weird to me on a conceptual level. So I'd say it is purely for aesthetic reason and feel free to change if it makes the server processor implementation simpler.
